packages/headless/src/ssr-next/commerce/controller-utils.test.ts

```diff
@@ -0,0 +1,276 @@
+          search: false,
+        });
+        controller2 = defineMockCommerceController({
+          search: false,
+        });
+
+        buildControllerDefinitions({
+          definitionsMap: {controller1, controller2},
+          engine: mockEngine,
+          solutionType: SolutionType.search,
+        });
+
+        expect(ControllerBuilder).not.toHaveBeenCalled();
+      });
+    });
+  });
+
+  describe('#createStaticState', () => {
+    beforeEach(() => {
+      const mockStaticBuilder = {
+        withState: vi.fn().mockReturnThis(),
+        withKind: vi.fn().mockReturnThis(),
+        withInitialState: vi.fn().mockReturnThis(),
+        build: vi.fn().mockReturnValue({
+          state: {},
+        }),
+      };
+
+      vi.mocked(createStaticControllerBuilder).mockReturnValue(
+        mockStaticBuilder
+      );
+
+      const controllers = {
+        controller1: buildMockController(),
+        controller2: buildMockControllerWithInitialState(mockEngine, {
+          initialState: {prop1: 'value1', prop2: 42},
+        }),
+      };
+
+      createStaticState({
+        searchActions: mockSearchActions,
+        controllers,
+      });
+    });
+
+    it('should call #createStaticControllerBuilder as many times as there are controllers', () => {
+      expect(createStaticControllerBuilder).toHaveBeenCalledTimes(2);
+    });
+
+    it('should call #createStaticControllerBuilder with the correct arguments', () => {
+      expect(createStaticControllerBuilder).toHaveBeenNthCalledWith(
+        1,
+        expect.objectContaining({
+          state: {},
+        })
+      );
+
+      expect(createStaticControllerBuilder).toHaveBeenNthCalledWith(
+        2,
+        expect.objectContaining({
+          state: {prop1: 'value1', prop2: 42},
+        })
+      );
+    });
+
+    it('should call #clone for each search action with the proper arguments', () => {
+      expect(utils.clone).toHaveBeenCalledTimes(2);
+      expect(utils.clone).toHaveBeenNthCalledWith(1, mockSearchActions[0]);
+      expect(utils.clone).toHaveBeenNthCalledWith(2, mockSearchActions[1]);
+    });
+  });
+
+  describe('#ensureAtLeastOneSolutionType', () => {
+    it('should not throw when no options provided', () => {
+      expect(() => ensureAtLeastOneSolutionType()).not.toThrow();
+    });
+
+    it('should not throw when listing is true', () => {
+      expect(() =>
+        ensureAtLeastOneSolutionType({listing: true, search: false})
+      ).not.toThrow();
+    });
+
+    it('should not throw when search is true', () => {
+      expect(() =>
+        ensureAtLeastOneSolutionType({listing: false, search: true})
+      ).not.toThrow();
+    });
+
+    it('should not throw when both listing and search are true', () => {
+      expect(() =>
+        ensureAtLeastOneSolutionType({listing: true, search: true})
+      ).not.toThrow();
+    });
+
+    it('should throw InvalidControllerDefinition when both listing and search are false', () => {
```

Comment:
Wouldn't it work with `standalone: true`, though? We're not really covering everything here.